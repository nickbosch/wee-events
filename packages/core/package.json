{
  "name": "@weegigs/events-core",
  "version": "0.16.1",
  "description": "Core Event System",
  "author": "Kevin O'Neill <kevin@weegigs.com.au>",
  "license": "MIT",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "packageManager": "yarn@3.1.1",
  "repository": {
    "type": "git",
    "url": "https://github.com/weegigs/wee-events.git"
  },
  "scripts": {
    "build": "yarn ws:build",
    "clean": "yarn ws:clean",
    "compile": "yarn ws:compile",
    "test": "yarn ws:test",
    "lint": "yarn ws:lint",
    "doctor": "yarn ws:doctor",
    "fix": "yarn run ws:fix",
    "publish": "yarn run ws:publish"
  },
  "dependencies": {
    "lodash": "^4.17.21",
    "luxon": "^2.4.0",
    "nanoid": "^3.3.4",
    "ulid": "^2.3.0",
    "zod": "^3.16.1"
  },
  "devDependencies": {
    "@tsconfig/node14": "^1.0.1",
    "@types/jest": "^27.5.1",
    "@types/lodash": "^4.14.182",
    "@types/luxon": "^2.3.2",
    "@types/node": "^14.18.18",
    "jest": "^27.5.1",
    "ts-jest": "^27.1.5",
    "typescript": "^4.6.4"
  },
  "files": [
    "lib"
  ],
  "jest": {
    "preset": "ts-jest",
    "testEnvironment": "node",
    "collectCoverageFrom": [
      "**/*.{ts,tsx}",
      "!**/node_modules/**"
    ],
    "testMatch": [
      "**/*.spec.ts",
      "**/*.test.ts"
    ],
    "collectCoverage": false,
    "coverageReporters": [
      "html"
    ]
  },
  "prettier": {
    "printWidth": 120,
    "trailingComma": "es5"
  }
}
