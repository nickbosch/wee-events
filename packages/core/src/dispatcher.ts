import { AggregateId, Command, Entity, Payload } from "./types";

export interface Dispatcher<S extends Payload, C extends Command> {
  execute(aggregate: AggregateId, command: C): Promise<Entity<S> | undefined>;
}
