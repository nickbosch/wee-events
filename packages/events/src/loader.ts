import { AggregateId, Entity, Payload } from "./types";

export interface Loader<S extends Payload> {
  load(aggregate: AggregateId): Promise<Entity<S> | undefined>;
}
