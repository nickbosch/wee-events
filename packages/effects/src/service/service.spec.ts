import _ from "lodash";

import * as T from "@effect-ts/core/Effect";
import { pipe } from "@effect-ts/core/Function";

import * as wee from "@weegigs/events-core";

import * as dispatcher from "./dispatcher";
import * as loader from "./loader";
import * as store from "../event-store";

import * as receipts from "./sample/receipts";

const id = { key: "test", type: "receipt" };
const ms = new wee.MemoryStore();

describe("describe receipt service", () => {
  beforeEach(() => ms.clear());

  it("should fail if entity can't be created", async () => {
    const program = pipe(receipts.service.load(id), T.provideService(store.EventLoader)(ms));

    expect(T.runPromise(program)).rejects.toBeInstanceOf(loader.EntityNotAvailableError);
  });

  it("should load an entity", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const program = pipe(receipts.service.load(id), T.provideService(store.EventLoader)(ms));
    const result = await T.runPromise(program);

    expect(result.state).toEqual({ total: 10 });
    expect(result.revision).not.toEqual(wee.Revision.Initial);
    expect(_.omit(result, "revision")).toMatchSnapshot();
  });

  it("should reject unknown paths", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("add/it", id, { amount: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    expect(T.runPromise(program)).rejects.toBeInstanceOf(dispatcher.HandlerNotFound);
  });

  it("should validate command structure", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("add", id, { value: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    expect(T.runPromise(program)).rejects.toBeInstanceOf(dispatcher.CommandValidationError);
  });

  it("should update an entity", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("add", id, { amount: 5 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));
    const result = await T.runPromise(program);

    expect(result.state).toEqual({ total: 15 });
    expect(result.revision).not.toEqual(wee.Revision.Initial);
    expect(_.omit(result, "revision")).toMatchSnapshot();
  });

  it("should propagate handler errors", async () => {
    ms.publish(id, { type: "added", data: { amount: 10 } });

    const update = receipts.service.execute("deduct", id, { amount: 15 });

    const program = pipe(update, T.provideService(store.EventLoader)(ms), T.provideService(store.EventPublisher)(ms));

    expect(T.runPromise(program)).rejects.toBeInstanceOf(receipts.InsufficientBalanceError);
  });
});
