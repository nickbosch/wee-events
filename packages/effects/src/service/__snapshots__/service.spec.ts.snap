// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe receipt service should load an entity 1`] = `
Object {
  "aggregate": Object {
    "key": "test",
    "type": "receipt",
  },
  "state": Object {
    "total": 10,
  },
  "type": "receipt",
}
`;

exports[`describe receipt service should update an entity 1`] = `
Object {
  "aggregate": Object {
    "key": "test",
    "type": "receipt",
  },
  "state": Object {
    "total": 15,
  },
  "type": "receipt",
}
`;
